swap/src/protocol/alice/event_loop.rs

```diff
@@ -103,59 +98,52 @@ pub struct EventLoop {
-    let peer_id = PeerId::from(identity.public());
-
-    let mut swarm = libp2p::swarm::SwarmBuilder::new(transport, behaviour, peer_id)
-        .executor(Box::new(TokioExecutor {
-            handle: tokio::runtime::Handle::current(),
-        }))
-        .build();
-
-    Swarm::listen_on(&mut swarm, listen_address.clone())
-        .with_context(|| format!("Address is not supported: {:#}", listen_address))?;
-
-    let recv_encrypted_signature = BroadcastChannels::default();
-    let send_transfer_proof = MpscChannels::default();
-
-    let handle_clone = EventLoopHandle {
-        recv_encrypted_signature: recv_encrypted_signature.sender.subscribe(),
-        send_transfer_proof: send_transfer_proof.sender.clone(),
-    };
-
-    let driver = EventLoop {
-        swarm,
-        peer_id,
-        execution_params,
-        bitcoin_wallet,
-        monero_wallet,
-        db,
-        listen_address,
-        swap_amounts: Default::default(),
-        recv_encrypted_signature: recv_encrypted_signature.sender,
-        send_transfer_proof: send_transfer_proof.receiver,
-        handle: handle_clone,
-    };
-
-    let handle = EventLoopHandle {
-        recv_encrypted_signature: recv_encrypted_signature.receiver,
-        send_transfer_proof: send_transfer_proof.sender,
-    };
-
-    Ok((driver, handle))
-}
-
 impl EventLoop {
+    pub fn new(
+        listen_address: Multiaddr,
+        seed: Seed,
+        execution_params: ExecutionParams,
+        bitcoin_wallet: Arc<bitcoin::Wallet>,
+        monero_wallet: Arc<monero::Wallet>,
+        db: Arc<Database>,
+    ) -> Result<Self> {
+        let identity = network::Seed::new(seed).derive_libp2p_identity();
+        let behaviour = Behaviour::default();
+        let transport = transport::build(&identity)?;
+        let peer_id = PeerId::from(identity.public());
+
+        let mut swarm = libp2p::swarm::SwarmBuilder::new(transport, behaviour, peer_id)
+            .executor(Box::new(TokioExecutor {
+                handle: tokio::runtime::Handle::current(),
+            }))
+            .build();
+
+        Swarm::listen_on(&mut swarm, listen_address.clone())
+            .with_context(|| format!("Address is not supported: {:#}", listen_address))?;
+
+        let recv_encrypted_signature = BroadcastChannels::default();
+        let send_transfer_proof = MpscChannels::default();
+
+        let handle = EventLoopHandle {
+            recv_encrypted_signature: recv_encrypted_signature.receiver,
+            send_transfer_proof: send_transfer_proof.sender.clone(),
+        };
+
+        Ok(Self {
+            swarm,
+            peer_id,
+            execution_params,
+            bitcoin_wallet,
+            monero_wallet,
+            db,
+            listen_address,
+            swap_amounts: Default::default(),
+            recv_encrypted_signature: recv_encrypted_signature.sender,
+            send_transfer_proof: send_transfer_proof.receiver,
+            handle,
```

Comment:
It feels weird that the `EventLoop` has a handle to itself.

Reply:
Actually we don't need the full handle but we do have to keep the `sender` end of the `mpsc::Channel` channels to be able to clone them when producing a new handle.

I'll remove the handle but it means the event loop will hold both ends of some channels.